gusto/equations.py

```diff
@@ -0,0 +1,168 @@
+from abc import ABCMeta, abstractproperty
+from collections import OrderedDict
+from firedrake import TestFunction, TrialFunction, MixedFunctionSpace, \
+    FiniteElement
+from gusto.diagnostics import Diagnostics
+from gusto.state import FieldCreator
+from gusto.terms import *
+from gusto.transport_terms import *
+
+
+class Equation(object):
+    """
+    Base equation class.
+
+    Creates test and trial functions on the given function space,
+    defines the mass term and provides a method for adding terms.
+    :arg function_space: :class:`.FunctionSpace` object. The function
+    space that the equation is defined on.
+
+    """
+    def __init__(self, function_space):
+        self.terms = OrderedDict()
+        self.test = TestFunction(function_space)
+        self.trial = TrialFunction(function_space)
+        self.bcs = None
+
+    def mass_term(self, q):
+        return inner(self.test, q)*dx
+
+    def add_term(self, term):
+        key = term.__class__.__name__
+        self.terms[key] = term
```

Comment:
This is a classic case where you should overload intrinsic operators to minimise notational clutter. Instead of calling this method `add_term`, you should call it `__iadd__`. This will cause the notation `equation += term` to work.

Reply:
Don't do that latter thing, since it swiftly becomes a nightmare.  Is it `(term, "label")` or `("label", term)`. What if you want to add some further thing? Why not instead just label the terms themselves, e.g., `foo = SomeTerm(); foo = foo.label("implicit")`

Where `foo.label` takes a shallow copy and adds a label (so that you don't have worries about sharing data).

Reply:
No, you can't add terms to magic methods. However I also think you shouldn't. The Right Thing is to build mathematical objects and apply mathematical operators to them. Terms are objects. If you want a label on the term then there should be an operator which takes a term and returns a term with a label. The labelled term should then be added to the equation. This way we keep composability and echo the maths more closely. 

I think Gusto is in a little danger of digging itself into a hole here. Can we get everyone relevant around a whiteboard and work out what the maths of this (and therefore the code) should look like?